File: loop_transformations/sources/unroll.4.f90
! @@name:       unroll.4.f
! @@type:       F-free
! @@compilable: yes
! @@linkable:   no
! @@expect:     success
! @@version:    omp_5.1
subroutine unroll_partial_remainder(n, A)
   implicit none
   integer :: n, i
   integer :: A(*)

    !$omp parallel do
    !$omp unroll partial(4)
    do i = 1, n
       A(i) = i
    end do

end subroutine

subroutine unroll_partial_remainder_option1(n, A)
   implicit none
   integer :: n, i_iv
   integer :: A(*)

   !$omp parallel do
   do i_iv = 0,(n+3)/4 -1
                               A(i_iv * 4 + 1) = i_iv * 4 + 1
        if (i_iv * 4 + 2 <= n) A(i_iv * 4 + 2) = i_iv * 4 + 2
        if (i_iv * 4 + 3 <= n) A(i_iv * 4 + 3) = i_iv * 4 + 3
        if (i_iv * 4 + 4 <= n) A(i_iv * 4 + 4) = i_iv * 4 + 4
   end do

end subroutine

subroutine unroll_partial_remainder_option2(n, A)
   implicit none
   integer :: n, i_iv, i_rem
   integer :: A(*)

    !$omp parallel do
    do i_iv = 0, (n+3)/4 -1
        if (i_iv < n/4) then
            A(i_iv * 4 + 1) = i_iv * 4 + 1
            A(i_iv * 4 + 2) = i_iv * 4 + 2
            A(i_iv * 4 + 3) = i_iv * 4 + 3
            A(i_iv * 4 + 4) = i_iv * 4 + 4
        else
           !! remainder loop
           do i_rem = i_iv*4 +1, n
              A(i_rem) = i_rem
           end do
        end if
    end do

end subroutine

subroutine unroll_partial_remainder_option3(n, A)
   implicit none
   integer :: n, i_iv, i_rem
   integer :: A(*)

    !$omp parallel do
    do i_iv = 0, (n/4) -1

       A(i_iv * 4 + 1) = i_iv * 4 + 1
       A(i_iv * 4 + 2) = i_iv * 4 + 2
       A(i_iv * 4 + 3) = i_iv * 4 + 3
       A(i_iv * 4 + 4) = i_iv * 4 + 4
    end do

    !! remainder loop
    !$omp parallel do
    do i_rem = (n/4)*4 +1, n
       A(i_rem) = i_rem
    end do

end subroutine

program main
implicit none
integer, parameter :: NT=12

integer :: i
logical :: error=.false.
integer   :: A(NT), C(NT)=[ (i, i=1,NT) ]

    A(1:NT)=0
    call unroll_partial_remainder(NT, A)
    if( .not. all(A(1:NT) == C(1:NT)) ) error = .true.

    A(1:NT)=0
    call unroll_partial_remainder_option1(NT, A)
    if( .not. all(A(1:NT) == C(1:NT)) ) error = .true.

    A(1:NT)=0
    call unroll_partial_remainder_option2(NT, A)
    if( .not. all(A(1:NT) == C(1:NT)) ) error = .true.

    A(1:NT)=0
    call unroll_partial_remainder_option3(NT, A)
    if( .not. all(A(1:NT) == C(1:NT)) ) error = .true.

    if(.not. error) print*, "OUT: Passed."
    if(      error) print*, "OUT: Failed"
end program

File: loop_transformations/sources/unroll.4.c
/*
* @@name:	unroll.4
* @@type:	C
* @@compilable:	yes
* @@linkable:	no
* @@expect:	success
* @@version:    omp_5.1
*/
void unroll_partial_remainder(int n, int A[])
{
    #pragma omp parallel for
    #pragma omp unroll partial(4)
    for (int i = 0; i < n; ++i)
        A[i] = i;
}

void unroll_partial_remainder_option1(int n, int A[])
{
    #pragma omp parallel for
    for (int i_iv = 0; i_iv < (n+3)/4; ++i_iv) {
                              A[i_iv * 4 + 0] = i_iv * 4 + 0;
        if (i_iv * 4 + 1 < n) A[i_iv * 4 + 1] = i_iv * 4 + 1;
        if (i_iv * 4 + 2 < n) A[i_iv * 4 + 2] = i_iv * 4 + 2;
        if (i_iv * 4 + 3 < n) A[i_iv * 4 + 3] = i_iv * 4 + 3;
   }
}

void unroll_partial_remainder_option2(int n, int A[])
{
    #pragma omp parallel for
    for (int i_iv = 0; i_iv < (n+3)/4; ++i_iv) {
        if (i_iv < n/4) {
            A[i_iv * 4 + 0] = i_iv * 4 + 0;
            A[i_iv * 4 + 1] = i_iv * 4 + 1;
            A[i_iv * 4 + 2] = i_iv * 4 + 2;
            A[i_iv * 4 + 3] = i_iv * 4 + 3;
        } else {
            // remainder loop
            for (int i_rem = i_iv*4; i_rem < n; ++i_rem)
                A[i_rem] = i_rem;
        }
    }
}

void unroll_partial_remainder_option3(int n, int A[])
{
    // main loop
    #pragma omp parallel for
    for (int i_iv = 0; i_iv < n/4; ++i_iv) {
        A[i_iv * 4 + 0] = i_iv * 4 + 0;
        A[i_iv * 4 + 1] = i_iv * 4 + 1;
        A[i_iv * 4 + 2] = i_iv * 4 + 2;
        A[i_iv * 4 + 3] = i_iv * 4 + 3;
    }

    // remainder loop
    #pragma omp parallel for
    for (int i_rem = (n/4)*4; i_rem < n; ++i_rem)
        A[i_rem] = i_rem;
}

#include <stdio.h>
#define NT 12

int main(){
int error=0, A[NT],C[NT];
for(int i = 0; i<NT; i++){ A[i]=0; C[i]=i; }

    for(int i = 0; i<NT; i++) A[i]=0.0;
    unroll_partial_remainder(NT,A);
    for(int i = 0; i<NT; i++) if(A[i] != C[i]) error=1;

    for(int i = 0; i<NT; i++) A[i]=0.0;
    unroll_partial_remainder_option1(NT,A);
    for(int i = 0; i<NT; i++) if(A[i] != C[i]) error=1;

    for(int i = 0; i<NT; i++) A[i]=0.0;
    unroll_partial_remainder_option2(NT,A);
    for(int i = 0; i<NT; i++) if(A[i] != C[i]) error=1;

    for(int i = 0; i<NT; i++) A[i]=0.0;
    unroll_partial_remainder_option3(NT,A);
    for(int i = 0; i<NT; i++) if(A[i] != C[i]) error=1;

  if(!error) printf("OUT: Passed\n");
  if( error) printf("OUT: Failed\n");
}
